### ISSUE TITLE
As a new user, I want to login with my Twitch account, so that I can access my dashboard

### ISSUE THREAD

**Issue by trezy** [trezy-studios/Trezy-VTT#1]
## Acceptance Criteria

* The home page has a "Login" button
* Clicking on the "Login" button prompts me to login with Twitch
* After completing the login with Twitch, I am taken to my dashboard

## Notes

Auth is going to be weird. The workflow is going to look something like this:
* On the client, have the user login to Twitch
* Send the new auth token to our API
* In the API, validate the Twitch auth token against the Twitch API
* Generate a custom Firebase Auth token using the user's Twitch user ID
* Store the Twitch auth token in Firestore with the Twitch user ID
* Send the new Firebase token to the client
* Back on the client, log the user into Firebase with the new auth token

The reason we want to generate two tokens is that the Firebase token will allow us to use the `auth` object in our security rules, which is a critical piece of securing any Firebase setup. We'll store the Twitch token in our database as well so we can revalidate it with every request.

## Sequence Diagram

![eNp9UE1PAyEQve-vmKMmetCklz00Iask3gxFPbPspCVt2QpD_fsCti6sGy 9UdTMfDbUQhBrtfr0G22TcmM1oRwpvP_WJdTeacXrp6lHmUWdy1IJCCs9f6](https://user-images.githubusercontent.com/442980/109390987-576e9280-78da-11eb-9dee-266df5fe6b75.png)

## References

* [Twitch Docs: Getting an Access Token](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/#oauth-authorization-code-flow)
* [Firebase Docs: Creating Custom Auth Tokens](https://firebase.google.com/docs/auth/admin/create-custom-tokens)
* [Firebase Docs: Using Custom Auth Tokens](https://firebase.google.com/docs/auth/web/custom-auth)

**Comment by Marenthyu** [comment]
TL;DR: Use Authroization Code Flow instead of Implicit Flow.

Looking at the linked document of the Twitch API and the flow described, you are indicating that we are intending to use the Implicit Auth Flow.

I don't think that is the way we should go here, as this means the token will be on the client side and in-transit to our backend server, making it theoretically accessible to MITM. (Unlikely thanks to HTTPS, but still.)

Additionally, the Implicit Flow does not give us a refresh Token to transparently renew the users' authentication via Twitch once their Authentication runs out, requiring them to re-authenticate through Twitch (... ignoring the fact that this is only a quick redirect loop if they didnt unauthorize us, but still a short disruption and full reload of the page for the end user.)

Implicit Auth is meant for purely client-side applications, but the User in our usecase does not have a need to call protected Endpoints of the Twitch API or log in to Chat from the client side. We're working fully Server side with Twitch.

Thus, I recommend we slightly modify the suggested Flow to utilize the [Authorization Code Flow](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/#oauth-authorization-code-flow) instead. That way, the Twitch Token does not fall into the users' hand and is not exposed to be stolen by, for example, a rogue Extension. The one Secret that authenticates them with our service is plenty enough.

So, to summarize the changes:
In (2), we receive an Authentication Code (instead of Token), which gets sent via (3) and in (4), the Code is exchanged for a Token, which can be sure to be valid as it was just created.

Additionally, we will receive a Refresh Token, which should be stored alongside the Twitch Token inside the Firestore, similar to (7).
